### PR TITLE
Restore original gDirectory after event loop

### PR DESCRIPTION
**Pull request by vepadulano**
In ROOT,`gDirectory` is a global variable that points to the currently open file in the program. Whenever a method such as `Write()` is called, e.g. when writing an histogram object in a file, the gDirectory will implicitly tell where that file is, without the need of specifying the file path when saving/writing. 

Within a distributed event loop, the input files of the RDataFrame are opened to retrieve the information about their clusters. It so happens that whenever a new call to `TFile` or similar is issued, the `gDirectory` variable will be changed to point to the last file opened. This happens both in C++ and in Python. 

The `TDirectory::TContext` class serves the purpose of storing the current `gDirectory` when instantiated, then restoring it when destroyed. This means that no matter how many times the `gDirectory` has been changed, it will be restored to its initial value if a `TContext` was set. 

In the `Dist` class the `gDirectory` is changed in the scope of the `get_clusters()` method. After the reduce phase of the process, if a call to `Write()` or similar was issued, then it won't correctly save the ROOT object to the initial file because the `gDirectory` was reset in the event loop. 

In C++ this could be avoided by explicitly creating a `TContext` at the beginning of the program and then calling its destructor at the end. In Python, a simple `del TContext` wouldn't guarantee a call to the C++ destructor. Instead, a context manager would enable a better management of the creation and the destruction of the `TContext`. Fortunately, PyROOT offers a pythonization that enables the call of the C++ destructor of any ROOT object through the `__destruct__()` dunder method. 

To summarise:
* A new factory function named `managed_tcontext` has been added in the `Proxy` module. This function is decorated with `contextlib.contextmanager` to make it work in a Python context manager without the need for a full class with `__enter__` and `__exit__` methods. This function creates a TContext at the beginning of the `with` statement and finally destroys it when exiting the context. 
* A new test has been added. It creates a file with some data and an empty one. The data is loaded in an RDataFrame and then an histogram is created from it. The histogram finally gets written to the empty file, then retrieved back from it to check that it was correctly saved.

### CODE REVIEW

File: tests/integration/spark/test_histo_write.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+        # Write histogram to out_file.root and close the file
+        histo.Write()
+        outfile.Close()
+
+        # Reopen file to check that histogram was correctly stored
+        reopen_file = ROOT.TFile("out_file.root", "read")
+        reopen_histo = reopen_file.Get("x")
+
+        # Check histogram statistics
+        self.assertEqual(reopen_histo.GetEntries(), 10000)
```
Comment (JavierCVilla) [inline tests/integration/spark/test_histo_write.py]: What do you think about using a Class attribute for the values `10000`, `10`, `1` and  `0.01`? IMO it may help to improve the readability and also to avoid changing the values in one function but not the other one (`create_tree_with_data`).

**Review by JavierCVilla — COMMENTED** [review]
Bravo @vepadulano, very elegant solution. 

I just have one minor comment, let you decide if you think it is worth to modify.

File: tests/integration/spark/test_histo_write.py (hunk 1 above)
Comment (vepadulano) [reply tests/integration/spark/test_histo_write.py]: yeah sure!
